idb/postgres/postgres_migrations.go

```diff
@@ -790,97 +686,372 @@ func m5RewardsAndDatesPart2UpdateAccounts(db *IndexerDb, state *MigrationState,
-		err = executeForEachCreatable(setCreateCloseAppLocal, address[:], result.appLocal)
+		// Don't update special accounts (m10 fixes this)
+		if (address != specialAccounts.FeeSink) && (address != specialAccounts.RewardsPool) {
+			if _, ok := accountsFirstUsed[address]; !ok {
+				accountData := initM7AccountData()
+
+				accountData.account.createdValid = true
+				accountData.account.created = 0
+				accountData.account.deletedValid = true
+				accountData.account.deleted = false
+
+				res = append(res, addressAccountData{address, accountData})
+			}
+		}
+
+		numRows++
+		if numRows%1000000 == 0 {
+			db.log.Printf("m7: read %d accounts", numRows)
+		}
+	}
+	db.log.Print("m7: finished reading accounts")
+
+	return res, nil
+}
+
+func updateAccounts(db *IndexerDb, maxRound uint32, specialAccounts idb.SpecialAccounts, accountsFirstUsed map[sdk_types.Address]txnID, readyAccountData []addressAccountData) error {
+	// Query all transactions again.
+	batchSize := 500
+
+	accountDataMap := make(map[sdk_types.Address]*m7AccountData)
+	assetDataMap := make(map[uint32]createClose)
+
+	numAccounts := len(accountsFirstUsed)
+	numAccountsUpdated := 0
+
+	db.log.Print("m7: querying transactions (pass 2)")
+	query := "SELECT round, intra, txnbytes, asset FROM txn WHERE round <= $1 ORDER BY round DESC, intra DESC"
+	rows, err := db.db.Query(query, maxRound)
+	if err != nil {
+		return fmt.Errorf("m7: unable to query transactions: %v", err)
+	}
+	defer rows.Close()
+
+	var writeDuration time.Duration = 0
+
+	// Loop through all transactions and compute account data.
+	db.log.Print("m7: started reading transactions (pass 2)")
+	numRows := 0
+	numBatches := 0
+	for rows.Next() {
+		var round uint32
+		var intra uint32
+		var txnBytes []byte
+		var assetID uint32
+		err = rows.Scan(&round, &intra, &txnBytes, &assetID)
+		if err != nil {
+			return fmt.Errorf("m7: unable to scan a row: %v", err)
+		}
+
+		var stxn types.SignedTxnWithAD
+		err = msgpack.Decode(txnBytes, &stxn)
 		if err != nil {
-			return fmt.Errorf("%s: failed to update %s with app local create/close: %v", rewardsCreateCloseUpdateErr, addressStr, err)
+			return fmt.Errorf("m7: unable to parse txnBytes round: %d intra: %d err: %v",
+				round, intra, err)
+		}
+
+		participants := getParticipants(stxn)
+		for _, address := range participants {
+			// Don't update special accounts (m10 fixes this).
+			if (address != specialAccounts.RewardsPool) && (address != specialAccounts.FeeSink) {
+				accountData, ok := accountDataMap[address]
+				if !ok {
+					accountData = initM7AccountData()
+					accountDataMap[address] = accountData
+				}
+
+				updateAccountData(address, round, assetID, stxn, accountData, &assetDataMap)
+
+				firstUsed, ok := accountsFirstUsed[address]
+				if !ok {
+					return fmt.Errorf("m7: accountFirstUsed does not contain address: %v", address)
+				}
+				if (txnID{uint32(round), uint32(intra)}) == firstUsed {
```

Comment:
I think we could get away with only storing round, it would require changing the flush condition to only flush when the round changes.

This implementation is really nice, I was expecting the code to be a lot more complicated.